---
# 📝 Taller 1: Preprocesamiento de textos & exploracion y extraccion de elementos de lenguaje
---
## Ingegrantes
*   Francisco Martinez Duque
*   Juan Diego Gomez Chavarro



In [386]:
!pip install -q datasets

In [387]:
from datasets import load_dataset

ds = load_dataset("jjiguaran/tweets_emotions_elections_colombia")

---
# 1️⃣ Normalicen el formato de sus textos:
---



In [388]:
print(ds.data)

{'train': MemoryMappedTable
text: string
tweet_id: int64
labels: string
labels_encoded: string
----
text: [["#NosUnimosONosJodemos #NosUnimosONosJodemos  Vamos juntos por el cambio, mire el miedo que le tienen.  @petrogustavo Mi presidente","@tobonsanin Otra vez este viejo estúpido politiquero, estamos es llevados cuando creemos  viejo decrépito de estos va hacer la solución para Colombia, ayudando a escoger corruptos como #quintero en MEDELLIN #PetroElPatronDelMal #FedericoEsColombia #NosUnimosONosJodemos","Sin importar encuestas como en el plebiscito juntos por Colombia el 29 nos vemos #FedericoImparable #NosUnimosONosJodemos #FedericoEsColombia","@FranciaMarquezM Ahora q es a ud si le parece grave? Y cuando sus borregos de la primera linea lo hacian contra nuestros policias q? Chevere cuando es para otros pero grave cuando es hacia mi? 🤥 #PetroElPatronDelMal #NosUnimosONosJodemos","Este es el Presidente que nos merecemos, recorrió el país entero, escuchó, expresó cariño y nos contag

In [389]:
# Se pasan todas los tweets a minusculas.
tweets = []
tweetsMinusculas = []

for textTweet in ds['train']['text']:
  tweets.append(textTweet)
  tweetsMinusculas.append(textTweet.lower())

In [390]:
!pip install -q emoji

## 🔹 Eliminar y reemplazar caracteres especiales, símbolos y emojis.

In [391]:
import re

def extraer_emojis(texto):
    return re.findall(r'[\U0001F300-\U0001FAD6]', texto)

emojisConjunto=set()

for texto in tweetsMinusculas:
  emojis = extraer_emojis(texto)
  for i in emojis:
    emojisConjunto.add(i)

In [392]:
emojis_significado = {
    '🤥': 'Mentiroso',
    '🦵': 'Pierna',
    '🔄': 'Actualizar',
    '🤡': 'Payaso',
    '🚨': 'Alerta',
    '💖': 'Corazón brillante',
    '🪠': 'Desatascador',
    '😂': 'Risa',
    '🤩': 'Asombro',
    '🥴': 'Mareado',
    '👀': 'Mirando',
    '💚': 'Corazón verde',
    '🏻': 'Tono de piel claro',
    '🎊': 'Confeti',
    '🤞': 'Suerte',
    '🛵': 'Motocicleta',
    '😐': 'Neutral',
    '🐊': 'Cocodrilo',
    '🕊': 'Paloma',
    '🤓': 'Nerd',
    '🏽': 'Tono de piel medio',
    '💃': 'Baile',
    '🤭': 'Sorpresa',
    '🤨': 'Escepticismo',
    '👉': 'Señalar',
    '🧐': 'Curioso',
    '😜': 'Broma',
    '🤘': 'Rock',
    '👹': 'Ogro',
    '😘': 'Beso',
    '💜': 'Corazón morado',
    '🎉': 'Celebración',
    '💙': 'Corazón azul',
    '🤯': 'Explosión mental',
    '🙏': 'Rezo',
    '👏': 'Aplausos',
    '🥳': 'Fiesta',
    '😥': 'Preocupado',
    '🤦': 'Frustración',
    '😏': 'Sarcástico',
    '🤣': 'Carcajada',
    '😡': 'Enojado',
    '🍳': 'Huevo frito',
    '🤐': 'Boca cerrada',
    '🏼': 'Tono de piel medio-claro',
    '👍': 'Pulgar arriba',
    '🍆': 'Berenjena',
    '🙄': 'Ojos en blanco',
    '🔎': 'Lupa',
    '🤑': 'Dinero',
    '🔵': 'Círculo azul',
    '🧑': 'Persona',
    '😅': 'Nervioso',
    '🎶': 'Música',
    '📱': 'Teléfono',
    '🔝': 'Top',
    '💸': 'Dinero volando',
    '👋': 'Saludo',
    '🔴': 'Círculo rojo',
    '😔': 'Triste',
    '🙃': 'Irónico',
    '🥲': 'Lágrima de alegría',
    '🥰': 'Enamorado',
    '💦': 'Gotas de agua',
    '🍗': 'Pollo',
    '🔃': 'Repetir',
    '🔥': 'Fuego',
    '🤮': 'Vómito',
    '🕺': 'Bailarín',
    '👇': 'Apuntar abajo',
    '💀': 'Calavera',
    '😉': 'Guiño',
    '🤔': 'Pensativo',
    '😎': 'Cool',
    '🗳': 'Voto',
    '🕳': 'Agujero',
    '👎': 'Pulgar abajo',
    '🏾': 'Tono de piel oscuro',
    '🙌': 'Victoria',
    '💔': 'Corazón roto',
    '😌': 'Aliviado',
    '😃': 'Feliz',
    '🥑': 'Aguacate',
    '😭': 'Llorando',
    '💩': 'Popó',
    '😞': 'Decepción',
    '😬': 'Incómodo',
    '🟡': 'Círculo amarillo',
    '💛': 'Corazón amarillo',
    '🪢': 'Nudo',
    '🐋': 'Ballena',
    '🥺': 'Suplicante',
    '🙋': 'Mano levantada',
    '😤': 'Molesto',
    '🐝': 'Abeja',
    '🤢': 'Asco',
    '💪': 'Fuerza',
    '🤗': 'Abrazo',
    '😍': 'Enamorado'
}


In [393]:
def reemplazar_emojis(texto):
    return re.sub(r'[\U0001F300-\U0001FAFF]', lambda m: emojis_significado.get(m.group(0), m.group(0)) + ' ', texto).strip()

for i in range(len(tweetsMinusculas)):
  tweetsMinusculas[i] = reemplazar_emojis(tweetsMinusculas[i])

---
# 2️⃣ Segmenten sus documentos a nivel de frase y palabra.
---

In [394]:
!pip install -q stanza

In [395]:
import stanza

stanza.download("es", verbose=False)

In [396]:
nlp_stanza = stanza.Pipeline(lang="es", processors="tokenize, lemma", verbose=False)

In [397]:
tweetsPalabras=dict()
tweetsFrases=dict()
for i in range(len(tweetsMinusculas)):
  doc_stanza=nlp_stanza(tweetsMinusculas[i])
  tokens_stanza_Palabra = [word.text for sentence in doc_stanza.sentences for word in sentence.words]
  tweetsPalabras[i]=tokens_stanza_Palabra
  tokens_stanza_Frase = [sentence.text for sentence in doc_stanza.sentences]
  tweetsFrases[i]=tokens_stanza_Frase


In [398]:
# Frases
for i in range(10):
  print("Frase", i, tweetsFrases[i])

Frase 0 ['#nosunimosonosjodemos #nosunimosonosjodemos  vamos juntos por el cambio, mire el miedo que le tienen.', '@petrogustavo mi presidente']
Frase 1 ['@tobonsanin otra vez este viejo estúpido politiquero, estamos es llevados cuando creemos  viejo decrépito de estos va hacer la solución para colombia, ayudando a escoger corruptos como #quintero en medellin #petroelpatrondelmal #federicoescolombia #nosunimosonosjodemos']
Frase 2 ['sin importar encuestas como en el plebiscito juntos por colombia el 29 nos vemos #federicoimparable #nosunimosonosjodemos #federicoescolombia']
Frase 3 ['@franciamarquezm ahora q es a ud si le parece grave? y cuando sus borregos de la primera linea lo hacian contra nuestros policias q? chevere cuando es para otros pero grave cuando es hacia mi?', 'Mentiroso  #petroelpatrondelmal #nosunimosonosjodemos']
Frase 4 ['este es el presidente que nos merecemos, recorrió el país entero, escuchó, expresó cariño y nos contagió la esperanza #federicoescolombia']
Frase 5

In [399]:
# Palabras
for i in range(10):
  print("Palabras tweet", i, tweetsPalabras[i])

Palabras tweet 0 ['#nosunimosonosjodemos', '#nosunimosonosjodemos', 'vamos', 'juntos', 'por', 'el', 'cambio', ',', 'mire', 'el', 'miedo', 'que', 'le', 'tienen', '.', '@petrogustavo', 'mi', 'presidente']
Palabras tweet 1 ['@tobonsanin', 'otra', 'vez', 'este', 'viejo', 'estúpido', 'politiquero', ',', 'estamos', 'es', 'llevados', 'cuando', 'creemos', 'viejo', 'decrépito', 'de', 'estos', 'va', 'hacer', 'la', 'solución', 'para', 'colombia', ',', 'ayudando', 'a', 'escoger', 'corruptos', 'como', '#quintero', 'en', 'medellin', '#petroelpatrondelmal', '#federicoescolombia', '#nosunimosonosjodemos']
Palabras tweet 2 ['sin', 'importar', 'encuestas', 'como', 'en', 'el', 'plebiscito', 'juntos', 'por', 'colombia', 'el', '29', 'nos', 'vemos', '#federicoimparable', '#nosunimosonosjodemos', '#federicoescolombia']
Palabras tweet 3 ['@franciamarquezm', 'ahora', 'q', 'es', 'a', 'ud', 'si', 'le', 'parece', 'grave', '?', 'y', 'cuando', 'sus', 'borregos', 'de', 'la', 'primera', 'linea', 'lo', 'hacian', 'cont

---
# 3️⃣ Eliminen signos de puntuación y caracteres no alfabéticos.
---
Calculen el número de tokens y palabras ditintas en sus textos.

In [400]:
tweetsMinusculasSinCaracteres=dict()
tweetsFrasesSinCaracteres=dict()

In [401]:
import re

def borrarCaracteresEspeciales(text):
    return re.sub(r"[^\w@#\s]","",text)

for i in range(len(tweetsMinusculas)):
  tweetsMinusculasSinCaracteres[i] = borrarCaracteresEspeciales(tweetsMinusculas[i])

for i in range(len(tweetsFrases)):
  frases= []
  for j in range(len(tweetsFrases[i])):
    frases.append(borrarCaracteresEspeciales(tweetsFrases[i][j]))
  tweetsFrasesSinCaracteres[i] = frases

In [402]:
for i in range(10):
  print("Frases Tweet", i, "sin caracteres", tweetsFrasesSinCaracteres[i])

Frases Tweet 0 sin caracteres ['#nosunimosonosjodemos #nosunimosonosjodemos  vamos juntos por el cambio mire el miedo que le tienen', '@petrogustavo mi presidente']
Frases Tweet 1 sin caracteres ['@tobonsanin otra vez este viejo estúpido politiquero estamos es llevados cuando creemos  viejo decrépito de estos va hacer la solución para colombia ayudando a escoger corruptos como #quintero en medellin #petroelpatrondelmal #federicoescolombia #nosunimosonosjodemos']
Frases Tweet 2 sin caracteres ['sin importar encuestas como en el plebiscito juntos por colombia el 29 nos vemos #federicoimparable #nosunimosonosjodemos #federicoescolombia']
Frases Tweet 3 sin caracteres ['@franciamarquezm ahora q es a ud si le parece grave y cuando sus borregos de la primera linea lo hacian contra nuestros policias q chevere cuando es para otros pero grave cuando es hacia mi', 'Mentiroso  #petroelpatrondelmal #nosunimosonosjodemos']
Frases Tweet 4 sin caracteres ['este es el presidente que nos merecemos reco

In [403]:
for i in range(10):
  print("Tweet", i, "sin caracteres", tweetsMinusculasSinCaracteres[i])

Tweet 0 sin caracteres #nosunimosonosjodemos #nosunimosonosjodemos  vamos juntos por el cambio mire el miedo que le tienen  @petrogustavo mi presidente
Tweet 1 sin caracteres @tobonsanin otra vez este viejo estúpido politiquero estamos es llevados cuando creemos  viejo decrépito de estos va hacer la solución para colombia ayudando a escoger corruptos como #quintero en medellin #petroelpatrondelmal #federicoescolombia #nosunimosonosjodemos
Tweet 2 sin caracteres sin importar encuestas como en el plebiscito juntos por colombia el 29 nos vemos #federicoimparable #nosunimosonosjodemos #federicoescolombia
Tweet 3 sin caracteres @franciamarquezm ahora q es a ud si le parece grave y cuando sus borregos de la primera linea lo hacian contra nuestros policias q chevere cuando es para otros pero grave cuando es hacia mi Mentiroso  #petroelpatrondelmal #nosunimosonosjodemos
Tweet 4 sin caracteres este es el presidente que nos merecemos recorrió el país entero escuchó expresó cariño y nos contagió 

In [404]:
tokensSinCaracteresEspeciales=dict()
for i in range(len(tweetsMinusculasSinCaracteres)):
  doc_stanza=nlp_stanza(tweetsMinusculasSinCaracteres[i])
  tokens_stanza_Palabra = [word.text for sentence in doc_stanza.sentences for word in sentence.words]
  tokensSinCaracteresEspeciales[i]=tokens_stanza_Palabra

In [405]:
for i in range(10):
  print("Tokens tweet:", i, tokensSinCaracteresEspeciales[i])

Tokens tweet: 0 ['#nosunimosonosjodemos', '#nosunimosonosjodemos', 'vamos', 'juntos', 'por', 'el', 'cambio', 'mire', 'el', 'miedo', 'que', 'le', 'tienen', '@petrogustavo', 'mi', 'presidente']
Tokens tweet: 1 ['@tobonsanin', 'otra', 'vez', 'este', 'viejo', 'estúpido', 'politiquero', 'estamos', 'es', 'llevados', 'cuando', 'creemos', 'viejo', 'decrépito', 'de', 'estos', 'va', 'hacer', 'la', 'solución', 'para', 'colombia', 'ayudando', 'a', 'escoger', 'corruptos', 'como', '#quintero', 'en', 'medellin', '#petroelpatrondelmal', '#federicoescolombia', '#nosunimosonosjodemos']
Tokens tweet: 2 ['sin', 'importar', 'encuestas', 'como', 'en', 'el', 'plebiscito', 'juntos', 'por', 'colombia', 'el', '29', 'nos', 'vemos', '#federicoimparable', '#nosunimosonosjodemos', '#federicoescolombia']
Tokens tweet: 3 ['@franciamarquezm', 'ahora', 'q', 'es', 'a', 'ud', 'si', 'le', 'parece', 'grave', 'y', 'cuando', 'sus', 'borregos', 'de', 'la', 'primera', 'linea', 'lo', 'hacian', 'contra', 'nuestros', 'policias', 

In [406]:
contadorTokensSinCaracteresEspeciales=0
palabrasUnicasSinCaracteresEspeciales=set()

for i in tokensSinCaracteresEspeciales:
  for j in tokensSinCaracteresEspeciales[i]:
    contadorTokensSinCaracteresEspeciales+=1
    palabrasUnicasSinCaracteresEspeciales.add(j)

print("Numero de tokens", contadorTokensSinCaracteresEspeciales)
print("Numero de palabras unicas", len(palabrasUnicasSinCaracteresEspeciales))

Numero de tokens 29117
Numero de palabras unicas 5770


Nº Tokens sin caracteres especiales = 29117

Nº Palabras del vocavulario sin caracteres especiales = 5770

Diferencias=23347

Quiere decir que existen 23347 palabras repetidas en todo el corpus, mientras que las palabras únicas son solo 5770

las palabras únicas representan el 19,8% del corpus.

---
# 4️⃣ Seleccionen un listado de posibles stop-words y eliminen esas palabras de sus textos.
---

In [407]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

nltk.download('stopwords', quiet=True)
nltk.download('punkt_tab', quiet=True)

True

In [408]:
stopwords_es = set(stopwords.words('spanish'))
print("Numero de stopwords:", len(stopwords_es))
for i in range(10):
  print("Stopword",i, ":", list(stopwords_es)[i])

print("...")

Numero de stopwords: 313
Stopword 0 : porque
Stopword 1 : estemos
Stopword 2 : había
Stopword 3 : su
Stopword 4 : hubiésemos
Stopword 5 : sentida
Stopword 6 : estos
Stopword 7 : nuestro
Stopword 8 : mío
Stopword 9 : hubiéramos
...


In [409]:
tokensSinStopWords = dict()
frasesSinStopWords = dict()

In [410]:
for i in range(len(tweetsMinusculasSinCaracteres)):
    tokensNLTKSinStopWords = [t for t in tokensSinCaracteresEspeciales[i] if t.lower() not in stopwords_es]
    frasesSinStopWords[i] = [" ".join(tokensNLTKSinStopWords)]
    tokensSinStopWords[i]=tokensNLTKSinStopWords

In [411]:
for i in range(10):
  print("Tweet", i, "sin Stop Words", frasesSinStopWords[i])

Tweet 0 sin Stop Words ['#nosunimosonosjodemos #nosunimosonosjodemos vamos juntos cambio mire miedo @petrogustavo presidente']
Tweet 1 sin Stop Words ['@tobonsanin vez viejo estúpido politiquero llevados creemos viejo decrépito va hacer solución colombia ayudando escoger corruptos #quintero medellin #petroelpatrondelmal #federicoescolombia #nosunimosonosjodemos']
Tweet 2 sin Stop Words ['importar encuestas plebiscito juntos colombia 29 vemos #federicoimparable #nosunimosonosjodemos #federicoescolombia']
Tweet 3 sin Stop Words ['@franciamarquezm ahora q ud si parece grave borregos primera linea hacian policias q chevere grave hacia Mentiroso #petroelpatrondelmal #nosunimosonosjodemos']
Tweet 4 sin Stop Words ['presidente merecemos recorrió país entero escuchó expresó cariño contagió esperanza #federicoescolombia']
Tweet 5 sin Stop Words ['familia democracia libertades #federicoescolombia']
Tweet 6 sin Stop Words ['bien colombia #federicoescolombia']
Tweet 7 sin Stop Words ['fico desarro

In [412]:
contadorTokensSinStopWords=0
for i in tokensSinStopWords:
  for j in tokensSinStopWords[i]:
    contadorTokensSinStopWords+=1

print("Numero de tokens sin contar las stopword:", contadorTokensSinStopWords)

Numero de tokens sin contar las stopword: 15759


In [413]:
palabrasUnicasSinStopWords=set()
for i in tokensSinStopWords:
  for j in tokensSinStopWords[i]:
    palabrasUnicasSinStopWords.add(j)

print("Numero de palabras sin contar las stopwords: ",len(palabrasUnicasSinStopWords))

Numero de palabras sin contar las stopwords:  5593


Nº Tokens sin caracteres especiales = 29117

Nº Tokens sin Stop Words = 15759

Diferencias en Tokens=13358

Nº Palabras del vocavulario sin caracteres especiales = 5770

Nº Palabras del vocavulario sin Stop Words = 5593

Diferencias en Palabras del vocavulario=177


---
# 5️⃣ Realicen tareas de stemming y lematizado en sus textos. De nuevo, recalculen el numero de tokens y el vocabulario. Comparen estos numeros con los obtenidos en los pasos anteriores.
---



## 🔹 Lemma

In [414]:
tokensLemma=dict()

In [415]:
for i in range(len(tokensSinStopWords)):
  tweetsTokensLematizados = []
  for j in range(len(tokensSinStopWords[i])):
    doc_stanza = nlp_stanza(tokensSinStopWords[i][j])
    lemma = doc_stanza.sentences[0].words[0].lemma
    tweetsTokensLematizados.append(lemma)
  tokensLemma[i] = tweetsTokensLematizados

In [416]:
for i in range(10):
  print("Tokens tweet", i, "Lematizado", tokensLemma[i])

Tokens tweet 0 Lematizado ['#nosunimosonosjodemo', '#nosunimosonosjodemo', 'ir', 'junto', 'cambio', 'mirar', 'miedo', '@petrogustavo', 'presidente']
Tokens tweet 1 Lematizado ['@tobonsanin', 'vez', 'viejo', 'estúpido', 'politiquero', 'llevar', 'creer', 'viejo', 'decrépito', 'ir', 'hacer', 'solución', 'colombia', 'ayudar', 'escoger', 'corrupto', '#quintero', 'medellin', '#petroelpatrondelmal', '#federicoescolombia', '#nosunimosonosjodemo']
Tokens tweet 2 Lematizado ['importar', 'encuesta', 'plebiscito', 'junto', 'colombia', '29', 'ver', '#federicoimparable', '#nosunimosonosjodemo', '#federicoescolombia']
Tokens tweet 3 Lematizado ['@franciamarquezm', 'ahora', 'q', 'ud', 'si', 'parecer', 'grave', 'borrego', 'primero', 'linea', 'hacian', 'policía', 'q', 'chevere', 'grave', 'hacia', 'Mentiroso', '#petroelpatrondelmal', '#nosunimosonosjodemo']
Tokens tweet 4 Lematizado ['presidente', 'merecemo', 'recorrer', 'país', 'entero', 'escuchar', 'expresar', 'cariño', 'contagiar', 'esperanza', '#fede

In [417]:
contadorTokensLematizdos=0
for i in tokensLemma:
  for j in tokensLemma[i]:
    contadorTokensLematizdos+=1

print("Numero de tokens despues de lematizar: ", contadorTokensLematizdos)

Numero de tokens despues de lematizar:  15759


In [418]:
palabrasUnicasLematizados=set()
for i in tokensLemma:
  for j in tokensLemma[i]:
    palabrasUnicasLematizados.add(j)

print("Numero de palabras despues de lematizar", len(palabrasUnicasLematizados))

Numero de palabras despues de lematizar 4331


## 🔹 Stemming

In [419]:
from nltk.stem import PorterStemmer

In [420]:
tokensStemming=dict()

In [421]:
stemmer_porter = PorterStemmer()
for i in range(len(tokensSinStopWords)):
  tweetsTokensStemming = []
  for j in range(len(tokensSinStopWords[i])):
    stemm = stemmer_porter.stem(tokensSinStopWords[i][j])
    tweetsTokensStemming.append(stemm)
  tokensStemming[i] = tweetsTokensStemming

In [422]:
for i in range(10):
  print("Tokens tweet", i, "Stemming", tokensStemming[i])

Tokens tweet 0 Stemming ['#nosunimosonosjodemo', '#nosunimosonosjodemo', 'vamo', 'junto', 'cambio', 'mire', 'miedo', '@petrogustavo', 'president']
Tokens tweet 1 Stemming ['@tobonsanin', 'vez', 'viejo', 'estúpido', 'politiquero', 'llevado', 'creemo', 'viejo', 'decrépito', 'va', 'hacer', 'solución', 'colombia', 'ayudando', 'escog', 'corrupto', '#quintero', 'medellin', '#petroelpatrondelm', '#federicoescolombia', '#nosunimosonosjodemo']
Tokens tweet 2 Stemming ['importar', 'encuesta', 'plebiscito', 'junto', 'colombia', '29', 'vemo', '#federicoimpar', '#nosunimosonosjodemo', '#federicoescolombia']
Tokens tweet 3 Stemming ['@franciamarquezm', 'ahora', 'q', 'ud', 'si', 'parec', 'grave', 'borrego', 'primera', 'linea', 'hacian', 'policia', 'q', 'chever', 'grave', 'hacia', 'mentiroso', '#petroelpatrondelm', '#nosunimosonosjodemo']
Tokens tweet 4 Stemming ['president', 'merecemo', 'recorrió', 'paí', 'entero', 'escuchó', 'expresó', 'cariño', 'contagió', 'esperanza', '#federicoescolombia']
Tokens

In [423]:
contadorTokensStemming=0
for i in tokensStemming:
  for j in tokensStemming[i]:
    contadorTokensStemming+=1

print("Numero de tokens luego de la lematizacion: ", contadorTokensStemming)

Numero de tokens luego de la lematizacion:  15759


In [424]:
palabrasUnicasStemming=set()
for i in tokensStemming:
  for j in tokensStemming[i]:
    palabrasUnicasStemming.add(j)

print("Numero de palabras luega de lematizar: ", len(palabrasUnicasStemming))

Numero de palabras luega de lematizar:  5158


Nº Tokens sin caracteres especiales = 29117

Nº Tokens sin Stop Words = 15759

Nº Tokens Lematizando = 15759

Nº Tokens Stemming = 15759

Nº Palabras del vocavulario sin caracteres especiales = 5770

Nº Palabras del vocavulario sin Stop Words = 5593

Nº Palabras del vocavulario Lematizando = 4331

Nº Palabras del vocavulario Stemming = 5158



---
# 6️⃣ Etiquetado gramatical (POS Tagging). Empleen una librería de su eleccion (por ejemplo, spaCy, Stanza, NLTK o CORE NLP) para etiquetar gramaticalmente cada token de sus textos.
---



Etiquetado gramatical

In [425]:
!python -m spacy download es_core_news_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.9/12.9 MB 95.4 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [426]:
import spacy
nlp_es = spacy.load("es_core_news_sm")

In [427]:
tokensEtiquetados = dict()

for i in range(len(tweetsMinusculasSinCaracteres)):
  tweetsTokensEtiquetados = []
  doc_es = nlp_es(tweetsMinusculasSinCaracteres[i])
  for token in doc_es:
    tweetsTokensEtiquetados.append((token.text, token.pos_))
  tokensEtiquetados[i] = tweetsTokensEtiquetados

In [428]:
print(f"{'Token':<20} {'Etiqueta':<10}")  # Print headers
print("-" * 30)  # Print a separator line

for i in range(1):
  for j in range(len(tokensEtiquetados[i])):
    print(f"{tokensEtiquetados[i][j][0]:<20} {tokensEtiquetados[i][j][1]:<10}")

Token                Etiqueta  
------------------------------
#                    NOUN      
nosunimosonosjodemos NOUN      
#                    NOUN      
nosunimosonosjodemos VERB      
                     SPACE     
vamos                VERB      
juntos               ADJ       
por                  ADP       
el                   DET       
cambio               NOUN      
mire                 VERB      
el                   DET       
miedo                NOUN      
que                  PRON      
le                   PRON      
tienen               VERB      
                     SPACE     
@petrogustavo        NOUN      
mi                   DET       
presidente           NOUN      


## 🔹 Significado de las Etiquetas en spaCy

Estas etiquetas corresponden a las categorías gramaticales utilizadas por **spaCy** para etiquetar palabras en un texto.  

### Lista de etiquetas y su significado

- **`PUNCT` (Punctuation)** → Signos de puntuación (`. , ; : ? !`).
- **`NUM` (Numeral)** → Números (`2024`, `tres`, `primero`).
- **`CCONJ` (Coordinating Conjunction)** → Conjunciones coordinantes (`y`, `o`, `pero`).
- **`DET` (Determiner)** → Determinantes (`el`, `la`, `un`, `mi`).
- **`ADP` (Adposition - Prepositions and Postpositions)** → Preposiciones (`en`, `a`, `de`, `sobre`).
- **`AUX` (Auxiliary Verb)** → Verbos auxiliares (`haber`, `ser`, `estar` en ciertos contextos).
- **`PROPN` (Proper Noun)** → Nombres propios (`España`, `Juan`, `Amazon`).
- **`ADJ` (Adjective)** → Adjetivos (`grande`, `rápido`, `feliz`).
- **`ADV` (Adverb)** → Adverbios (`rápidamente`, `muy`, `ayer`).
- **`VERB` (Verb)** → Verbos (`correr`, `hablar`, `comer`).
- **`SPACE` (Space)** → Espacios en blanco (`\n`, ` `).
- **`PRON` (Pronoun)** → Pronombres (`yo`, `tú`, `él`, `nosotros`).
- **`SCONJ` (Subordinating Conjunction)** → Conjunciones subordinantes (`porque`, `aunque`, `si`).
- **`INTJ` (Interjection)** → Interjecciones (`¡ay!`, `¡eh!`, `¡oops!`).
- **`NOUN` (Noun)** → Sustantivos (`casa`, `perro`, `computadora`).
- **`PART` (Particle)** → Partículas gramaticales que no encajan en otras categorías (`no`, `que` en ciertas estructuras como "tengo que ir").



## 🔹 Comenten y analicen brevemente los resultados, haciendo enfasis en posibles casos de ambiguedad o inconsistencias en la asignacion de etiquetas.

* El etiquetado POS, tiene algunos errores al intentar clasificar cada palabra, posiblemente esto se debe a que el trata de encontrar una estructura definida en las frases del corpus, sin embargo algunos de los usuarios no escriben los tweets con una estructura definida, un ejemplo de ello es el primer tweet:

[('#', 'NOUN'), ('nosunimosonosjodemos', 'NOUN'), ('#', 'NOUN'), ('nosunimosonosjodemos', 'VERB'), (' ', 'SPACE'), ('vamos', 'VERB'), ('juntos', 'ADJ'), ('por', 'ADP'), ('el', 'DET'), ('cambio', 'NOUN'), ('mire', 'VERB'), ('el', 'DET'), ('miedo', 'NOUN'), ('que', 'PRON'), ('le', 'PRON'), ('tienen', 'VERB'), (' ', 'SPACE'), ('@petrogustavo', 'NOUN'), ('mi', 'DET'), ('presidente', 'NOUN')]

Donde inicialmente clasifica nosunimosonosjodemos como sujeto y luego lo clasifica como verbo.

---
# 7️⃣ Analisis de distribucion de etiquetas
---

In [429]:
etiquetasDiferentes = set()
for i in range(len(tokensEtiquetados)):
  for j in range(len(tokensEtiquetados[i])):
    etiquetasDiferentes.add(tokensEtiquetados[i][j][1])

In [430]:
print(etiquetasDiferentes)

{'CCONJ', 'PUNCT', 'PART', 'ADP', 'VERB', 'SPACE', 'PRON', 'NOUN', 'ADJ', 'SCONJ', 'INTJ', 'NUM', 'PROPN', 'DET', 'ADV', 'AUX'}


In [431]:
frecuenciaEtiquetas = {
    'PUNCT': 0,
    'NUM': 0,
    'CCONJ': 0,
    'DET': 0,
    'ADP': 0,
    'AUX': 0,
    'PROPN': 0,
    'ADJ': 0,
    'ADV': 0,
    'VERB': 0,
    'SPACE': 0,
    'PRON': 0,
    'SCONJ': 0,
    'INTJ': 0,
    'NOUN': 0,
    'PART': 0
}

for i in range(len(tokensEtiquetados)):
  for j in range(len(tokensEtiquetados[i])):
      frecuenciaEtiquetas[tokensEtiquetados[i][j][1]]+=1

In [432]:
print("Etiqueta".ljust(10), "Frecuencia")
print()
for i in (frecuenciaEtiquetas):
  print(f"{i:<10} {frecuenciaEtiquetas[i]}")

Etiqueta   Frecuencia

PUNCT      2
NUM        313
CCONJ      1040
DET        3651
ADP        3883
AUX        1287
PROPN      1579
ADJ        2717
ADV        1575
VERB       3842
SPACE      826
PRON       2150
SCONJ      911
INTJ       30
NOUN       7252
PART       7


## 🔹 Presenten esta informacion en una tabla o gráfica. Discutan qué patrones se observan y cómo podrían relacionarse con la naturaleza del corpus.

* Muchos tweets mencionan personas, marcas, lugares o eventos, lo que es coherente con el uso de Twitter para discusiones sobre figuras públicas y temas de tendencia.

* La baja presencia de signos de puntuación es debido al pre-procesamiento del corpus.

* Debido a la cantidad tan elevada de adjetivos (2717) se puede inferir que el contenido del corpus es muy crítico o esta basada en opiniones.

---
# 8️⃣ Analisis sintactico
---

In [433]:
from stanza.utils.visualization.dependency_visualization import visualize_strings

numeroTweet = [0 ,4, 6]

for i in numeroTweet:
  visualizarFrases = tweetsFrases[i]
  visualize_strings(visualizarFrases, "es")
  print(visualizarFrases)


['#nosunimosonosjodemos #nosunimosonosjodemos  vamos juntos por el cambio, mire el miedo que le tienen.', '@petrogustavo mi presidente']


['este es el presidente que nos merecemos, recorrió el país entero, escuchó, expresó cariño y nos contagió la esperanza #federicoescolombia']


['por el bien de colombia #federicoescolombia']


---
# 9️⃣ Evaluacion del analisis sintactico
---

* #nosunimosonosjodemos #nosunimosonosjodemos  vamos juntos por el cambio mire el miedo que le tienen  @petrogustavo mi presidente.

* este es el presidente que nos merecemos recorrió el país entero escuchó expresó cariño y nos contagió la esperanza #federicoescolombia.

* por el bien de colombia #federicoescolombia.


---
# 🔟 Identificacion de entidades nombradas
---

In [434]:
### Una prueba que vi en los cuadernos de colab de la gente de Stanford
from stanza.utils.visualization.ner_visualization import visualize_strings

es_strings = [tweets[1],
              tweets[200],
              tweets[250],
              tweets[300]]


visualize_strings(es_strings, "es")

/usr/local/lib/python3.11/dist-packages/spacy/displacy/__init__.py:213: UserWarning: [W006] No entities to visualize found in Doc object. If this is surprising to you, make sure the Doc was processed using a model that supports named entity recognition, and check the `doc.ents` property manually if necessary.
  warnings.warn(Warnings.W006)


## 🔹 Comenten brevemente los resultados y dificultades encontradas, especialmente en el caso de entidades ambiguas o poco frecuentes.

* El clasificador tiene algunos problemas en la identificación de las entidades nombradas con los caracteres especial del @. Aunque en algunos lo hace de manera correcta, en algunos casos no reconoce la entidad.

por ejemplo en el tweet:

@NoticiasCaracol Qué no se note el sesgo de @NestorMoralesC PER , no sorprende en un " periodista " pro Duque-Fico PER . #DebateFinal MISC

Reconoce la entidad @NestorMoralesC PER pero no reconoce @NoticiasCaracol.

* Por otro lado, también se presenta problemas con la etiqueta numeral #, por ejemplo en el primer tweet:

@tobonsanin Otra vez este viejo estúpido politiquero , estamos es llevados cuando creemos viejo decrépito de estos va hacer la solución para Colombia LOC , ayudando a escoger corruptos como #quintero MISC en MEDELLIN #PetroElPatronDelMal ORG # FedericoEsColombia MISC # NosUnimosONosJodemos MISC

la entidad no es reconocida en todos los casos con la etiqueta #.



---
# 1️⃣1️⃣ Evaluacion básica de NER. Seleccionen manualmente un subconjunto de sus documentos y comparen las entidades reconocidas por su sistema con las que realmente aparecen en el texto (gold standard o revision manual). Calculen metricas como precision, exhaustividad (recall) y F1.
---

In [435]:
es_strings = [tweets[0],
              tweets[1],
              tweets[2],
              tweets[3],
              tweets[4],
              tweets[5],
              tweets[6],
              tweets[7],
              tweets[8],
              tweets[9],
              tweets[10],
              tweets[11],
              tweets[12],
              tweets[13],
              tweets[14],
              tweets[15],
              tweets[16],
              tweets[17],
              tweets[18],
              tweets[19]]


visualize_strings(es_strings, "es")

## 🔹 Cálculo de métricas: Precisión, Recall y F1-score

Dado los siguientes datos:

- \( TN = 329 \) (Verdaderos Negativos)
- \( FP = 10 \) (Falsos Positivos)
- \( TP = 41 \) (Verdaderos Positivos)
- \( FN = 15 \) (Falsos Negativos)

#### **1. Precisión (Precision)**  
La precisión se define como:

$$
Precision = \frac{TP}{TP + FP}
$$

Sustituyendo los valores:

$$
Precision = \frac{41}{41 + 10} = \frac{41}{51} \approx 0.80
$$

#### **2. Exhaustividad (Recall o Sensibilidad)**  
El recall se define como:

$$
Recall = \frac{TP}{TP + FN}
$$

Sustituyendo los valores:

$$
Recall = \frac{41}{41 + 15} = \frac{41}{56} \approx 0.73
$$

#### **3. F1-score**  
El F1-score se define como:

$$
F1 = 2 \times \frac{Precision \times Recall}{Precision + Recall}
$$

Sustituyendo los valores calculados:

$$
F1 = 2 \times \frac{0.8039 \times 0.7321}{0.8039 + 0.7321}
$$

$$
F1 = 2 \times \frac{0.5883}{1.536} \approx 2 \times 0.3830 = 0.77
$$

#### **Resultados Finales:**
- **Precisión**: \( 0.80 \) (80%)
- **Recall**: \( 0.73 \) (73%)
- **F1-score**: \( 0.77 \) (77%)




---
# 1️⃣2️⃣ Agregar < s > y < /s > a cada frase
---

In [436]:
tweetsFrasesMarcas = dict()
tweetsFrasesSinMarcas = dict()
frasesSinMarcas = tweetsFrasesSinCaracteres
frasesSinMarcas = frasesSinStopWords

for i in range(len(frasesSinMarcas)):
  frases = []
  frases2 =[]
  for j in range(len(frasesSinMarcas[i])):
    frases.append("<s>"+frasesSinMarcas[i][j]+"</s>")
    frases2.append(frasesSinMarcas[i][j])
  tweetsFrasesMarcas[i] =frases
  tweetsFrasesSinMarcas[i] =frases2

In [437]:
for i in range(10):
  print("Frase tweet", i, tweetsFrasesMarcas[i])

Frase tweet 0 ['<s>#nosunimosonosjodemos #nosunimosonosjodemos vamos juntos cambio mire miedo @petrogustavo presidente</s>']
Frase tweet 1 ['<s>@tobonsanin vez viejo estúpido politiquero llevados creemos viejo decrépito va hacer solución colombia ayudando escoger corruptos #quintero medellin #petroelpatrondelmal #federicoescolombia #nosunimosonosjodemos</s>']
Frase tweet 2 ['<s>importar encuestas plebiscito juntos colombia 29 vemos #federicoimparable #nosunimosonosjodemos #federicoescolombia</s>']
Frase tweet 3 ['<s>@franciamarquezm ahora q ud si parece grave borregos primera linea hacian policias q chevere grave hacia Mentiroso #petroelpatrondelmal #nosunimosonosjodemos</s>']
Frase tweet 4 ['<s>presidente merecemos recorrió país entero escuchó expresó cariño contagió esperanza #federicoescolombia</s>']
Frase tweet 5 ['<s>familia democracia libertades #federicoescolombia</s>']
Frase tweet 6 ['<s>bien colombia #federicoescolombia</s>']
Frase tweet 7 ['<s>fico desarrollo fico progreso #f

## 🔹 Construccion de n-gramas

In [438]:
from nltk import ngrams, FreqDist

In [439]:
unigramas = []
bigramas = []
trigramas = []
for i in range(len(tweetsFrasesSinMarcas)):
  for j in range(len(tweetsFrasesSinMarcas[i])):
    doc_stanza=nlp_stanza(tweetsFrasesSinMarcas[i][j])
    tokens_stanza_Palabra = [word.text for sentence in doc_stanza.sentences for word in sentence.words]
    #tokens = word_tokenize(tweetsFrasesMarcas[i][j], language="spanish")  # Tokenización
    unigramas_frase = tokens_stanza_Palabra  # Generar unigramas
    bigramas_frase = list(ngrams(tokens_stanza_Palabra, 2))  # Generar bigramas
    trigramas_frase = list(ngrams(tokens_stanza_Palabra, 3))  # Generar trigramas

    for unigrama in unigramas_frase:
      unigramas.append(unigrama)
    for bigrama in bigramas_frase:
      bigramas.append(bigrama)
    for trigrama in trigramas_frase:
      trigramas.append(trigrama)

In [440]:
freq_unigramas = FreqDist(unigramas)
freq_bigramas = FreqDist(bigramas)
freq_trigramas = FreqDist(trigramas)

In [441]:
print(f"{'Palabra':<20} {'Frecuencia'}")
print("-"*33)
for i, (uni) in enumerate(freq_unigramas.items()):
    if i == 10:
        break
    print(f"{uni[0]:<20} {uni[1]}")

Palabra              Frecuencia
---------------------------------
#nosunimosonosjodemos 5
vamos                40
juntos               3
cambio               88
mire                 2
miedo                23
@petrogustavo        91
presidente           66
@tobonsanin          1
vez                  24


In [442]:
print(f"{'Bigrama':<30} {'Frecuencia':<10}")
print("-" * 55)  # Línea separadora

# Iterar con formato de columnas
for i, (bigram, count) in enumerate(freq_bigramas.items(), start=1):
  if (i == 10):
    break
  bigram_str = "-".join(bigram)
  # Formato con anchos fijos: <12 (izquierda), <25 (izquierda), <10 (izquierda)
  print(f"{bigram_str:<30} {count:<10}")

Bigrama                        Frecuencia
-------------------------------------------------------
#nosunimosonosjodemos-#nosunimosonosjodemos 1         
#nosunimosonosjodemos-vamos    1         
vamos-juntos                   2         
juntos-cambio                  1         
cambio-mire                    1         
mire-miedo                     1         
miedo-@petrogustavo            1         
@petrogustavo-presidente       4         
@tobonsanin-vez                1         


In [443]:
print(f"{'Trigrama':<55} {'Frecuencia':<10}")
print("-" * 65)  # Línea separadora

# Iterar con formato de columnas
for i, (tri, count) in enumerate(freq_trigramas.items(), start=1):
  if (i == 10):
    break
  tri_str = "--".join(tri)
  # Formato con anchos fijos: <12 (izquierda), <25 (izquierda), <10 (izquierda)
  print(f"{tri_str:<30} {count:<10}")

Trigrama                                                Frecuencia
-----------------------------------------------------------------
#nosunimosonosjodemos--#nosunimosonosjodemos--vamos 1         
#nosunimosonosjodemos--vamos--juntos 1         
vamos--juntos--cambio          1         
juntos--cambio--mire           1         
cambio--mire--miedo            1         
mire--miedo--@petrogustavo     1         
miedo--@petrogustavo--presidente 1         
@tobonsanin--vez--viejo        1         
vez--viejo--estúpido           1         


In [444]:

print(f"{'Unigrama':<30} {'Probabilidad':<10}")
probabilidadUnigrama = dict()
for i in range(len(unigramas)):
  probabilidadUnigrama[unigramas[i]] = freq_unigramas[unigramas[i]]/len(unigramas)
  if i <= 10:
    print(f"{unigramas[i]:<30} {probabilidadUnigrama[unigramas[i]]:<10.6f}")


#print(probabilidadUnigrama)


Unigrama                       Probabilidad
#nosunimosonosjodemos          0.000317  
#nosunimosonosjodemos          0.000317  
vamos                          0.002535  
juntos                         0.000190  
cambio                         0.005578  
mire                           0.000127  
miedo                          0.001458  
@petrogustavo                  0.005768  
presidente                     0.004183  
@tobonsanin                    0.000063  
vez                            0.001521  


In [445]:
probabilidadBigrama = dict()
print(f"{'Unigrama':<30} {'Probabilidad':<10}")
for i in range(len(bigramas)):
  probabilidadBigrama[bigramas[i]] = freq_bigramas[bigramas[i]]/freq_unigramas[bigramas[i][0]]
  if i <= 10:
    bigram_str = "-".join(bigramas[i])
    print(f"{bigram_str:<30} {probabilidadBigrama[bigramas[i]]:<10.6f}")


Unigrama                       Probabilidad
#nosunimosonosjodemos-#nosunimosonosjodemos 0.200000  
#nosunimosonosjodemos-vamos    0.200000  
vamos-juntos                   0.050000  
juntos-cambio                  0.333333  
cambio-mire                    0.011364  
mire-miedo                     0.500000  
miedo-@petrogustavo            0.043478  
@petrogustavo-presidente       0.043956  
@tobonsanin-vez                1.000000  
vez-viejo                      0.041667  
viejo-estúpido                 0.111111  


In [446]:
probabilidadTrigrama  = dict()

print(f"{'Unigrama':<50} {'Probabilidad':<10}")
for i in range(len(trigramas)):
  probabilidadTrigrama[trigramas[i]] = freq_trigramas[trigramas[i]]/freq_bigramas[trigramas[i][0:2]]
  if i <= 10:
    trigram_str = "-".join(trigramas[i])
    print(f"{trigram_str:<50} {probabilidadTrigrama[trigramas[i]]:<10.6f}")

Unigrama                                           Probabilidad
#nosunimosonosjodemos-#nosunimosonosjodemos-vamos  1.000000  
#nosunimosonosjodemos-vamos-juntos                 1.000000  
vamos-juntos-cambio                                0.500000  
juntos-cambio-mire                                 1.000000  
cambio-mire-miedo                                  1.000000  
mire-miedo-@petrogustavo                           1.000000  
miedo-@petrogustavo-presidente                     1.000000  
@tobonsanin-vez-viejo                              1.000000  
vez-viejo-estúpido                                 1.000000  
viejo-estúpido-politiquero                         1.000000  
estúpido-politiquero-llevados                      1.000000  


In [447]:
# Tamaño del vocabulario
#tokensUnigrama = len(set(unigramas))

# Función para calcular probabilidades SIN suavizado Laplace
#def ngram_prob(bigram, freq_uni, freq_bi):
#    previous = bigram[0]
#    return (freq_bi[bigram]) / (freq_uni[previous])

# Función para calcular probabilidades con suavizado Laplace
#def laplace_smoothed_prob(bigram, freq_uni, freq_bi, V):
#    previous = bigram[0]
#    return (freq_bi[bigram] + 1) / (freq_uni[previous] + V)

---
# 1️⃣3️⃣ Colocaciones
---

In [448]:
import math
def informacionMutua(freq_bigramas, freq_unigramas, bigrama):
    palabra1, palabra2 = bigrama
    frecuencia_bigrama = freq_bigramas[bigrama]
    frecuencia_palabra1 = freq_unigramas[palabra1]
    frecuencia_palabra2 = freq_unigramas[palabra2]
    probabilidad_bigrama = frecuencia_bigrama / len(bigramas)
    probabilidad_palabra1 = frecuencia_palabra1 / len(unigramas)
    probabilidad_palabra2 = frecuencia_palabra2 / len(unigramas)

    informacion_mutua = math.log2(probabilidad_bigrama / (probabilidad_palabra1 * probabilidad_palabra2))
    return informacion_mutua



In [449]:
informacionMutuaBigrama =  dict()
for bigrama in bigramas:
  informacionMutuaBigrama[bigrama] = informacionMutua(freq_bigramas, freq_unigramas, bigrama)

In [450]:
for i, (bi) in enumerate(informacionMutuaBigrama):
  if (i<14):
    print(i, bi, informacionMutuaBigrama[bi])

0 ('#nosunimosonosjodemos', '#nosunimosonosjodemos') 9.41570918954365
1 ('#nosunimosonosjodemos', 'vamos') 6.41570918954365
2 ('vamos', 'juntos') 8.152674783709857
3 ('juntos', 'cambio') 6.015171259959922
4 ('cambio', 'mire') 6.600133760681078
5 ('mire', 'miedo') 8.536003423261361
6 ('miedo', '@petrogustavo') 3.0282087830626656
7 ('@petrogustavo', 'presidente') 3.5073766197612253
8 ('@tobonsanin', 'vez') 9.474602878597219
9 ('vez', 'viejo') 6.304677877154907
10 ('viejo', 'estúpido') 10.889640377876063
11 ('estúpido', 'politiquero') 13.059565379318375
12 ('politiquero', 'llevados') 13.059565379318375
13 ('llevados', 'creemos') 13.059565379318375


---
# 1️⃣4️⃣ Generador de frases
---

## 🔹 Generador utilizando n-grams

In [451]:
probabilidadUnigramaOrdenada = dict(sorted(probabilidadUnigrama.items(), key=lambda x: x[1], reverse=True))
probabilidadBigramaOrdenada = dict(sorted(probabilidadBigrama.items(), key=lambda x: x[1], reverse=True))
probabilidadTrigramaOrdenada = dict(sorted(probabilidadTrigrama.items(), key=lambda x: x[1], reverse=True))


In [452]:
lambda_1 = 0.1
lambda_2 = 0.1
lambda_3 = 0.8

def probabilidadSiguientePalabra(palabraGenerada, ultimaPalabra, penultimaPalabra):
  p1 = lambda_1 * probabilidadUnigrama[palabraGenerada]

  if ((ultimaPalabra, palabraGenerada) not in probabilidadBigrama or ultimaPalabra == None):
    p2 = 0
  else:
    p2 = lambda_2 * probabilidadBigrama[(ultimaPalabra, palabraGenerada)]

  if ((penultimaPalabra, ultimaPalabra, palabraGenerada) not in probabilidadTrigrama or ultimaPalabra == None):
    p3 = 0
  else:
    p3 = lambda_3 * probabilidadTrigrama[(penultimaPalabra, ultimaPalabra, palabraGenerada)]

  return p1+p2+p3

def palabraMasProbable(ultimaPalabra, penultimaPalabra):
  probabilidadMaxima = 0
  palabra = ""
  for i in unigramas:
    if (probabilidadSiguientePalabra(i, ultimaPalabra, penultimaPalabra) >= probabilidadMaxima):
      probabilidadMaxima = probabilidadSiguientePalabra(i, ultimaPalabra, penultimaPalabra)
      palabra = i
    palabra
  return palabra

import random
def generarPrimeraPalabraAleatoria():
  return random.choice(list(probabilidadUnigramaOrdenada.keys()))

In [453]:


for j in range(5):
  primeraPalabra = ""
  frase = ""
  for i in range(20):
    segundoPalabra = generarPrimeraPalabraAleatoria()
    frase = frase + " " + primeraPalabra
    terceraPalabra = palabraMasProbable(primeraPalabra, segundoPalabra)
    primeraPalabra = segundoPalabra
    segundoPalabra = terceraPalabra

  print("<s>"+ frase+ "</s>")

<s>  corruptela huevo cárceles portal pgggesidente ciudadano votos buenoque comunes conformismo fanatismo hágame @supersalud juegan inquietud crecer acabé @vanedelatorre inmunidad</s>
<s>  apenas 2022 pendejada piensa pasaste 400 saca multiplicar tierras susto llegaremos #esmuyimportante protagonistas convencen estrategas manrique mimetiza rockero márquez</s>
<s>  minuto arriba vote dueles vi ven adulterado recorrió individual cuáles quieren bites conozca 14 35pm infamias tumbaron investigados reconoce</s>
<s>  azar pecho crea unas afirmación provocar allegados diplomáticas @ojodemocracia22 imparcial virgen puedo quiénes cambió andrea tituló #mujeres @antony_munoz iluminado</s>
<s>  favores relación leen pronunciar 5051 colonialistas manrique juro orador guerras mesías traten confundir descuido demasiado razón común cacaos amor</s>


## 🔹 Markovify

In [454]:
!pip install markovify

In [455]:
import markovify

In [456]:
corpus_es=""
for i in range(len(tweetsMinusculasSinCaracteres)):
    corpus_es = corpus_es + tweetsMinusculasSinCaracteres[i] + ".\n"

In [457]:

# Crear un solo modelo con state_size=2
modelo_es = markovify.Text(corpus_es, state_size=2)

# Generar 5 oraciones con parámetros para más variedad
print("=== Oraciones generadas en español ===")
for _ in range(5):
    oracion_es = modelo_es.make_sentence(tries=100, max_overlap_ratio=0.8, max_words=20)
    if oracion_es:
        print(f"Texto generado: {oracion_es}")
    else:
        print("No se pudo generar una oración válida (intenta con un corpus más grande).")

=== Oraciones generadas en español ===
Texto generado: #yaessuficiente que el programa de gobierno.
Texto generado: @petrogustavo ahora el culpado es quien nos va a dar toda en segunda #noalasmaquinariaspoliticas #segundavuelta #todosconrodolfohernandez.
Texto generado: #locorrectoes que va mal en las redes Escepticismo Molesto #lacorrupciónpetrista Lupa.
Texto generado: @noticiascaracol qué no va a venir a ponerle impuestos si escucho #eselmomentode.
Texto generado: #petropresidenteenprimera y dejo enlace de prueba gratis a mi familia y amigos es por la patria #voyconfederico.


---
# 1️⃣5️⃣ Repitan el paso 12 y calculen los modelos de lenguaje a nivel de bigrama (probabilidades de los bigramas),pero esta vez dejen un documento por fuera del corpus de entrada. Utilicen este documento para evaluar la calidad de su modelo de lenguaje empleando la medida de perplejidad.
---

In [458]:
import math

In [459]:
import random
def divCorpus (corpus, size=0.9):
  corpus_train = ""
  corpus_test = ""
  for i in range(len(corpus)):
    numero=random.uniform(0,1)
    if numero < size:
      corpus_train = corpus_train + corpus[i] + ".\n"
    else:
      corpus_test = corpus_test + corpus[i] + ".\n"
  return corpus_train, corpus_test


In [460]:
corpus_es=""
for i in range(len(tweetsMinusculasSinCaracteres)):
    corpus_es = corpus_es + tweetsMinusculasSinCaracteres[i] + ".\n"

In [461]:
corpus_train, corpus_test = divCorpus(tweetsMinusculasSinCaracteres)

In [462]:
from nltk import word_tokenize
from nltk import bigrams

In [463]:
import re
import math
from nltk import bigrams, FreqDist, word_tokenize
from nltk.corpus import stopwords

def metricas(textoTrain, textoTest):
    texto_prueba_es = textoTest
    tokens_prueba_es = texto_prueba_es.split()
    text_es = textoTrain

    # Función para limpiar y tokenizar el texto
    def tokenize_text(text, language='spanish'):
        tokens = word_tokenize(text)
        # Filtrar tokens alfanuméricos y pasar a minúsculas
        tokens = [token.lower() for token in tokens if re.match(r'\w+', token)]
        # Eliminar stop words
        stops = set(stopwords.words(language))
        tokens = [token for token in tokens if token not in stops]
        return tokens

    # Tokenización
    tokens_es = tokenize_text(text_es, language='spanish')

    # Construcción de bigramas
    bigrams_es = list(bigrams(tokens_es))

    # Frecuencias de unigramas y bigramas
    freq_uni_es = FreqDist(tokens_es)
    freq_bi_es = FreqDist(bigrams_es)

    # Tamaño del vocabulario
    V_es = len(set(tokens_es))

    # === Funciones de evaluación ===
    def compute_perplexity(tokens, freq_uni, freq_bi, V):
        N = len(tokens)
        log_prob_sum = 0

        for i in range(1, N):
            bg = (tokens[i-1], tokens[i])
            prob = (freq_bi.get(bg, 0) + 1) / (freq_uni.get(tokens[i-1], 0) + V)
            log_prob_sum += math.log(prob)

        return math.exp(-log_prob_sum / (N - 1))

    def compute_entropy(tokens, freq_uni, freq_bi, V):
        N = len(tokens)
        log_prob_sum = 0

        for i in range(1, N):
            bg = (tokens[i-1], tokens[i])
            prob = (freq_bi.get(bg, 0) + 1) / (freq_uni.get(tokens[i-1], 0) + V)
            log_prob_sum += math.log(prob)

        return -log_prob_sum / (N - 1)

    def compute_cross_entropy(tokens, freq_uni, freq_bi, V):
        N = len(tokens)
        log_prob_sum = 0

        for i in range(1, N):
            bg = (tokens[i-1], tokens[i])
            prob = (freq_bi.get(bg, 0) + 1) / (freq_uni.get(tokens[i-1], 0) + V)
            log_prob_sum += math.log(prob)

        return -log_prob_sum / (N - 1)

    # Evaluación en español
    print("\n=== Evaluación del modelo en español ===")
    perp_es = compute_perplexity(tokens_prueba_es, freq_uni_es, freq_bi_es, V_es)
    entropy_es = compute_entropy(tokens_prueba_es, freq_uni_es, freq_bi_es, V_es)
    cross_entropy_es = compute_cross_entropy(tokens_prueba_es, freq_uni_es, freq_bi_es, V_es)

    print(f"Perplejidad: {perp_es:.4f}")
    print(f"Entropía: {entropy_es:.4f} nats")
    print(f"Entropía cruzada: {cross_entropy_es:.4f} nats")

# === Ejemplo de evaluación ===
for i in range(10):
    Ntrain=random.uniform(0,1)
    corpus_train, corpus_test = divCorpus(tweetsMinusculasSinCaracteres, Ntrain)
    metricas(corpus_train, corpus_test)
    print("Ntrain: ", Ntrain)




=== Evaluación del modelo en español ===
Perplejidad: 1575.0895
Entropía: 7.3621 nats
Entropía cruzada: 7.3621 nats
Ntrain:  0.16892442154853227

=== Evaluación del modelo en español ===
Perplejidad: 2369.6848
Entropía: 7.7705 nats
Entropía cruzada: 7.7705 nats
Ntrain:  0.2904392911954048

=== Evaluación del modelo en español ===
Perplejidad: 3690.6124
Entropía: 8.2135 nats
Entropía cruzada: 8.2135 nats
Ntrain:  0.5453332599257562

=== Evaluación del modelo en español ===
Perplejidad: 4423.3759
Entropía: 8.3947 nats
Entropía cruzada: 8.3947 nats
Ntrain:  0.7379047633895213

=== Evaluación del modelo en español ===
Perplejidad: 3441.9430
Entropía: 8.1438 nats
Entropía cruzada: 8.1438 nats
Ntrain:  0.545759971599009

=== Evaluación del modelo en español ===
Perplejidad: 2754.0628
Entropía: 7.9208 nats
Entropía cruzada: 7.9208 nats
Ntrain:  0.35537017684801553

=== Evaluación del modelo en español ===
Perplejidad: 1686.1649
Entropía: 7.4302 nats
Entropía cruzada: 7.4302 nats
Ntrain:  0.1

## Conclusion:

* Cuando se emplea una muestra de entrenamiento más pequeña que la muestra de prueba, se observa que la perplejidad es más baja. Esto es posible porque, al ser más diversa la muestra de prueba, es mucho más probable encontrar frases parecidas a las frases entrenadas.

* Esto indica que el corpus de entrenamiento no es completamente representativo; se debe tener un corpus mucho más extenso y diverso.

---
# 1️⃣6️⃣- Implementen un modulo de corrección ortográfica para su sistema.
---

In [464]:
!pip install -q language-tool-python

In [466]:
import language_tool_python
tool = language_tool_python.LanguageTool('es')

In [467]:
listaPalabras = list(palabrasUnicasSinStopWords)
print("Palabra incorrecta   sugerencia")
for i in range(len(listaPalabras)):
  palabra=listaPalabras[i]
  correcciones=tool.check(palabra)
  if (i==100):
    break
  if correcciones:
    print(f"Palabra: {palabra}",end="")
    for error in correcciones:
        print(f" -----> {error.replacements}")

Palabra incorrecta   sugerencia
Palabra: rodolfohipocritas -----> []
Palabra: ingrid -----> ['Íngrid', 'Ingrid', 'Ingri', 'ingerid', 'ingrida', 'ingrido', 'injerid', 'In-Grid']
Palabra: comenzo -----> ['comenzó', 'comienzo', 'comentó', 'comento', 'comenzá']
Palabra: asi -----> ['así', 'si', 'casi', 'sí', 'ahí', 'Asia', 'as', 'Asís', 'asa', 'ANSI', 'ay', 'asir', 'asía', 'asó', 'asió', 'aso', 'asé', 'AI', 'API', 'ASIN', 'Aly', 'Alí', 'Amy', 'Ani', 'Asp', 'Asti', 'Asín', 'RSI', 'Yasy', 'acid', 'ají', 'ali', 'ami', 'ansí', 'ase', 'asid', 'asá', 'asín', 'asís', 'psi', 'tasi', 'a si', 'as i', 'AVI']
Palabra: caterba -----> ['caterva', 'cateaba', 'catera', 'catervas']
Palabra: anamaría -----> ['Anamaria', 'animaría', 'Anamari', 'acamaría', 'adamaría', 'afamaría', 'ana maría', 'Ana María']
Palabra: través -----> ['a través']
Palabra: venezuela -----> ['Venezuela', 'vejezuela', 'Cerezuela', 'bestezuela', 'bebésela', 'manezuela', 'netezuela', 'pernezuela', 'tenésela', 'vejezuelas', 'vejezuelo', 